lib/charliecloud.py

```diff
@@ -273,137 +229,108 @@ def fixup(self):
+
+         Empty layers are skipped.
+
+         Important note: TarFile.extractall() extracts the given members in
+         the order they are specified, so we need to preserve their order from
+         the file, as returned by getmembers(). We also need to quickly remove
+         members we don't want from this sequence. Thus, we use the OrderedSet
+         class defined in this module."""
       TT = collections.namedtuple("TT", ["fp", "members"])
-      if (self.layer_hashes is None):
-         self.layer_hashes_load()
       layers = collections.OrderedDict()
       # Schema version one (v1) allows one or more empty layers for Dockerfile
       # entries like CMD (https://github.com/containers/skopeo/issues/393).
       # Unpacking an empty layer doesn't accomplish anything so we ignore them.
       empty_cnt = 0
-      for (i, lh) in enumerate(self.layer_hashes, start=1):
-         INFO("layer %d/%d: %s: listing" % (i, len(self.layer_hashes), lh[:7]))
-         path = self.layer_path(lh)
+      for (i, path) in enumerate(layer_tars, start=1):
+         lh = os.path.basename(path).split(".", 1)[0]
+         lh_short = lh[:7]
+         INFO("layer %d/%d: %s: listing" % (i, len(layer_tars), lh_short))
          try:
             fp = TarFile.open(path)
-            members_list = fp.getmembers()  # reads whole file :(
+            members = OrderedSet(fp.getmembers())  # reads whole file :(
          except tarfile.TarError as x:
             FATAL("cannot open: %s: %s" % (path, x))
-         members = collections.OrderedDict([(m, None) for m in members_list])
          if (lh in layers and len(members) > 0):
-            FATAL("duplicate non-empty layer %s" % lh[:7])
+            FATAL("duplicate non-empty layer %s" % lh)
          if (len(members) > 0):
             layers[lh] = TT(fp, members)
          else:
             empty_cnt += 1
-      if (self.schema_version == '1'):
-         DEBUG('reversing layer order for schema version one (v1)')
-         layers = collections.OrderedDict(reversed(layers.items()))
       DEBUG("skipped %d empty layers" % empty_cnt)
       return layers
 
-   def pull_to_unpacked(self, use_cache=True, fixup=False, last_layer=None):
-      """Pull and flatten image. If fixup, then also add the Charliecloud
-         workarounds to the image directory."""
-      self.download(use_cache)
-      self.flatten(last_layer)
-      if (fixup):
-         self.fixup()
+   def tarballs_write(self, tarball_dir):
+      """Write one uncompressed tarball per layer to tarball_dir. Return a
+         sequence of tarball basenames, with the lowest layer first."""
+      # FIXME: Yes, there is only one layer for now and we'll need to update
+      # it when (if) we have multiple layers. But, I wanted the interface to
+      # support multiple layers.
+      base = "%s.tar" % self.ref.for_path
+      path = tarball_dir // base
+      try:
+         INFO("layer 1/1: gathering")
```

Comment:
In other places I see you use a for loop with an `enumerate`. Should we do that here to provide us with the `1/1` instead of this hardcoded print?

Reply:
I think things will probably change enough that any loop would need to be rewritten anyway when it's time.